src/data/helper/sourceManager.ts

```diff
@@ -346,8 +350,78 @@ export class SourceManager {
      * @param sourceIndex By defualt 0, means "main source".
      *                    Most cases there is only one source.
      */
-    getSource(sourceIndex?: number) {
-        return this._sourceList[sourceIndex || 0];
+    getSource(sourceIndex?: number): Source {
+        sourceIndex = sourceIndex || 0;
+        const source = this._sourceList[sourceIndex];
+        if (!source) {
+            // Series may share source instance with dataset.
+            const upSourceMgrList = this._getUpstreamSourceManagers();
+            return upSourceMgrList[0]
+                && upSourceMgrList[0].getSource(sourceIndex);
+        }
+        return source;
+    }
+
+    /**
+     * Will return undefined if source don't have dimensions.
+     *
+     * Only available for series.
+     */
+    getDataStorage(): DataStorage | undefined {
+        if (__DEV__) {
+            assert(isSeries(this._sourceHost), 'Can only call getDataStorage on series source manager.');
+        }
+        const source = this.getSource(0);
+        const dimensionsDefine = source.dimensionsDefine;
+
+        // Source from endpoint(usually series) will be read differently
+        // when seriesLayoutBy or startIndex(which is affected by sourceHeader) are different.
+        // So we use this three props as key.
+        const sourceReadKey = source.seriesLayoutBy
+            + '$$'
+            + source.startIndex
+            + (dimensionsDefine ? map(dimensionsDefine, def => def.name).join('$$') : '');
+
+        return this._innerGetDataStorage(source, sourceReadKey);
+    }
+
+    private _innerGetDataStorage(seriesSource: Source, sourceReadKey: string): DataStorage | undefined {
+        // TODO Can use other sourceIndex?
+        const sourceIndex = 0;
+
+        const storeList = this._storeList;
+
+        let cachedStoreMap = storeList[sourceIndex];
+
+        if (!cachedStoreMap) {
+            cachedStoreMap = storeList[sourceIndex] = {};
+        }
+
+        let cachedStore = cachedStoreMap[sourceReadKey];
+        if (!cachedStore) {
+            const upSourceMgr = this._getUpstreamSourceManagers()[0];
+
+            if (isSeries(this._sourceHost) && upSourceMgr) {
+                cachedStore = upSourceMgr._innerGetDataStorage(seriesSource, sourceReadKey);
+            }
+            else {
+                // Always create datastorage based on source from series.
+                const dimensionsDefine = seriesSource.dimensionsDefine;
+                // Can't create a store if don't know dimension..
+                if (seriesSource && dimensionsDefine) {
```

Comment:
In this case. Dataset will still generate `dimensionsDefine` for the source. `dimensionsDefine` is only undefined when the source is `series.data` instead of dataset. In which case sharing storage is meaningless.

Reply:
Use this option:
```js
option = {
    dataset: {
        source: [
            [12, 55, 33],
            [11, 85, 77]
        ]
    },
    xAxis: {},
    yAxis: {},
    series: [{
        type: 'scatter',
        encode: {x: 0, y: 1},
    }, {
        type: 'scatter',
        encode: {x: 0, y: 2},
    }]
};
```
`dimensionsDefine` is `undefined` here. But I've not dig into what's wrong with it yet.

Reply:
My mistake, seems dimensionDefine will be undefined if `dimensions` is not given.

Reply:
I add default dimensions generation in the dataset. It seems more reasonable than picking the dimensions from the coordinate system from the dataset side. But for better display in the tooltip. it will still use `displayName` from coordinate system. Which is 'x', 'y', 'value', etc.